File: python_translation/compute_graph_sl.py
import numpy as np

def compute_graph_sl(V, distance_vectors, weights, th=None):
    """
    Compute distance-based graph from Zhang et al., ICIP 2014.

    Parameters:
        V (numpy.ndarray): nx3 array. n points.
        distance_vectors (numpy.ndarray): Distance vectors.
        weights (numpy.ndarray): Weights.
        th (float): Threshold to construct the graph (optional).

    Returns:
        numpy.ndarray: Weight matrix representing the graph.
        numpy.ndarray: Edge list.
        numpy.ndarray: Degree vector.
        numpy.ndarray: Inverse distance matrix.
        numpy.ndarray: Indices of the closest points along the mean direction.
    """
    N = V.shape[0]

    if th is None:
        th = np.sqrt(3) + 0.00001

    mean_X = np.mean(V[:, 0])
    mean_Y = np.mean(V[:, 1])
    mean_Z = np.mean(V[:, 2])
    mean_point_cloud = np.array([mean_X, mean_Y, mean_Z])

    centered_V = V - mean_point_cloud

    mean_direction = np.sum(distance_vectors * weights) / np.sum(weights)
    mean_direction /= np.linalg.norm(mean_direction)

    squared_norms = np.sum(V**2, axis=1)
    D = np.sqrt(np.tile(squared_norms, (N, 1)) + np.tile(squared_norms[:, np.newaxis], (1, N)) - 2 * np.dot(V, V.T))
    iD = 1 / D
    iD[D > th] = 0
    iD[D == 0] = 0

    degrees = np.sum(iD, axis=1)
    degrees /= np.linalg.norm(degrees)

    sorted_indices = np.argsort(degrees)
    below_threshold = int(0.2 * len(sorted_indices))
    idx_closest_original = sorted_indices[:below_threshold]

    selected_vectors = centered_V[idx_closest_original]
    dot_products_degreed = np.dot(selected_vectors, mean_direction)

    idx_closest = np.argsort(dot_products_degreed)[:below_threshold]
    idx_closest = idx_closest_original[idx_closest[0]] if below_threshold > 0 else None

    W = iD.T + iD

    idx = np.nonzero(iD)
    I, J = np.unravel_index(idx, D.shape)
    edge = np.column_stack((I, J))

    return W, edge, degrees, iD, idx_closest

File: matlab_code/compute_graph_sl.m
function [W,edge,degrees,iD,idx_closest] = compute_graph_sl(V,distance_vector,weights, th)
% V: nx3. n points
% th: threshold to construct the graph
  N = size(V,1);
  %

  if(nargin==3)
      
      th =sqrt(3)+0.00001;
      
  end
  
% Calculate the mean along each dimension (X, Y, Z)
mean_X = mean(V(:, 1));
mean_Y = mean(V(:, 2));
mean_Z = mean(V(:, 3));

% Calculate the center (mean) of the point cloud
mean_point_cloud = [mean_X, mean_Y, mean_Z];

% Subtract the mean values from each corresponding dimension to center the point cloud
centered_V = V - mean_point_cloud;
  % Calculate dot products of points with the mean direction vector
mean_direction = sum(distance_vector .* weights) / sum(weights);  % Promedio ponderado según la intensidad del cambio de color
mean_direction = mean_direction/norm(mean_direction);     


%size(mean_direction)

% Count the number of values below the threshold
%below_threshold = round(0.2 * numel(sort(dot_products)))
%[~, sorted_indices] = sort(dot_products);
%idx_closest = sorted_indices(1:below_threshold);  % Segundo argumento threshold
% Find the index of the point closest and furthest along the mean direction   %Índices dentro del vector de Volumen del bloque
%[~, idx_furthest] = max(dot_products);

% Retrieve the points closest and furthest along the mean direction
  
  %compute EDM 
  squared_norms = sum(V.^2,2);
  D = sqrt(repmat(squared_norms,1,N) + repmat(squared_norms',N,1) - 2*(V*V'));
  % D = squareform(pdist(coords, 'euclidean')); % pairwise distances, n-by-n
  % matrix% only use pdist if have the statistics/ML toolbox
  % Formula tan conocida.
  iD = D.^(-1);                     % Inverso escalar de los elementos de la matriz, pesos es el inverso de la distancia
  iD(find(D > th)) = 0;             % Encontrar todas las distancias que sean mayores al umbral y evaluarlas en 0
  iD(find(D==0))   =0;              % Además encontrar aquellas que sean nulas, es decir, self-connections
  
  degrees = zeros(size(iD,1),1);
  for i = 1:size(iD,1)
      degrees(i) = sum(iD(:,i));
  end
  
  degrees = degrees/norm(degrees);  
  %dot_products_degreed = dot_products.*degrees;
  % Count the number of values below the threshold
  %below_threshold = sum(degrees < mean(degrees));
  %[sorted_degrees, sorted_indices] = sort(degrees);
  %below_threshold = round(0.2 * numel(sorted_degrees));
  
  [sorted_degrees, sorted_indices] = sort(degrees);
  below_threshold = round(0.2 * numel(sorted_degrees));

  idx_closest_original = sorted_indices(1:below_threshold);
 
% Select the corresponding vectors from centered_V using idx_closest_original
  selected_vectors = centered_V(idx_closest_original, :);

% Calculate dot products between selected vectors and the mean_direction
  dot_products_degreed = selected_vectors * mean_direction';

% Find the new indices of the closest points along the mean direction
  [~, idx_closest] = sort(dot_products_degreed);

% Use the new indices to reorder the original indices
  below_threshold = round(0.2 * numel(sort(dot_products_degreed)));
  if below_threshold > 0
    idx_closest = idx_closest_original(idx_closest(1));
  end  
  W=iD' + iD;                       % Matriz de adyacencia, agrega la transpuesta para hacer la matriz simétrica

  idx = find(iD~=0);                

  [I, J] = ind2sub( size(D), idx ); % Se identifica cuales son los nodos conectados

   edge = [I, J];
   



end